Adds residual chart (issue #486)

ds26gte opened the pull request:
This adds the function residual-chart-from-list that takes xs, ys, and residuals and draws a line plot with stick intervals from each y to its corresponding residual.

ds26gte committed:
- chart-lib.js: Added intervalChart() #486
- chart.art: Added residual-chart-from-list() (uses intervalChart)

schanzer commented:
@ds26gte this is really exciting! I tried the following program, and the resulting chart isn't quite what I expected.
```
include chart
xs  = [list: 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
ys  = [list: 2, 4, 6, 8,10,12,14,16,18, 20]
res = [list: 1, 2, 4,-1,-2,-4, 0, 4,-3,  3]

series = from-list.residual-chart(xs, ys, res)
c = render-chart(series)
c.display()
```
![image](https://github.com/brownplt/code.pyret.org/assets/77126/51b88304-25c4-4f73-ba0e-e34f00189ba6)

Some of those residuals should be on the other side of the line. Am I doing something wrong here?

schanzer commented on src/web/js/trove/chart-lib.js:
```diff
@@ -771,6 +772,35 @@
+
+    const options = {
+      curveType: 'function',
+      lineWidth: 2,
```
This should be 0, since we ONLY want to see the residuals.

shriram commented:
@schanzer Nothing wrong, I just think you've forgotten about a little thing called gravity.

shriram commented:
@schanzer Nothing wrong, I just think you've forgotten about a little thing called gravity.

schanzer commented:
Ah yes! The residuals are the y', not delta-y. Silly me 🤦‍♂️

schanzer commented:
@jpolitz I'll defer to you on the design of the code, but this raises no red flags for me. 

But in reading through this, I think we need a lower-level Pyret-facing API that includes some configurability so that we can use intervals in other areas as well. From reading [the documentation](https://developers.google.com/chart/interactive/docs/gallery/intervals), it looks like we want interval-chart to take in a list of xs, ys, a _list of lists_ (each of which represents an interval). I propose that it also expose a method `.type()`, which allows for stick, bar, box, and point intervals. I could also see a universe in which it exposes a `.color()` method, which colors the bars/sticks/boxes/points.

This would allow us to replicate most of the interval types in Google's API, though only one-at-a-time. But that's easily addressed by composing multiple interval charts view `render-charts`, so ultimately this would give us quite a bit of power.

The functionality of `residual-chart` could still exist, being built on top of this either as part of the chart library or implemented via a Bootstrap library.

ds26gte committed:
residual-chart-from-list: 3rd arg shd be offset, not y-prime #486

ds26gte commented:
> Ah yes! The residuals are the y', not delta-y. Silly me 🤦‍♂️

Corrected. _Residual_ does mean what you expected, so I've corrected the 3rd arg to be _Δy_ rather than _y'_. 

(Of course, please check that the direction of the offset is also as you would expect. I really don't know for sure.)

ds26gte committed:
Set interval-chart colors in chart.arr rather than chart-lib.js #486

ds26gte committed:
interval-chart: - point-size and lineWidth settable in chart.arr #486
- don't connect points as curve

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1046,6 +1046,8 @@ type IntervalChartSeries = {
   pointers :: Option<RawArray<Pointer>>,
   pointer-color :: Option<I.Color>,
+  point-size :: Number,
+  lineWidth :: Number,
```
For consistency's sake, shouldn't this be `line-width`?

ds26gte replied:
Unfortunately, the existing series use camelCase and kebab-case for different fields, and I was trying to be consistent with them.

schanzer replied:
makes sense. @jpolitz any interest in Dorai opening another PR in which these get normalized across the charting library?

jpolitz replied:
No issue with it, though having it separate is preferred.

jpolitz replied:
(Separate PR I mean, not separate naming conventions :-))

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1055,6 +1057,8 @@ default-interval-chart-series = {
   pointers: none,
   pointer-color: some(C.color(228, 147, 7, 1)),
+  point-size: 4,
+  lineWidth: 1,
```
`line-width`?

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1039,6 +1039,31 @@ default-bar-chart-series = {
+  tab :: TableIntern,
+  axisdata :: Option<AxisData>,
+  color :: Option<I.Color>,
+  colors :: Option<RawArray<I.Color>>,
```
What's the plural `colors` for here? I don't think it's used in the JS implementation

ds26gte replied:
In case we use multiple residuals?

schanzer replied:
Multiple residuals should be handled via multiple instances of `interval-chart`, composed via `render-charts(...)`

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1039,6 +1039,31 @@ default-bar-chart-series = {
+
+default-interval-chart-series = {
+  color: some(C.red),
+  colors: none,
```
same question - what are `colors` for here?

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1039,6 +1039,31 @@ default-bar-chart-series = {
+  pointer-color: some(C.color(228, 147, 7, 1)),
+  point-size: 4,
+  lineWidth: 1,
+  axisdata: none,
```
should this be `axis-data`, in keeping with Pyret conventions?

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
   data-series.make-axis(max-positive-height, max-negative-height)
 end
 
+fun residual-chart-from-list(
```
Can we rename this `interval-chart-from list`, in keeping with the data series renaming?

schanzer reviewed (changes requested):
@ds26gte left a few comments for you, but for the most part this looks _awesome!_

@jpolitz hoping you can weigh in on this as well.

ds26gte committed:
- Rename (from-list.)residual-chart to interval-chart #486
- Remove unused (yet) `colors` field from IntervalChartSeries

ds26gte committed:
chart.arr: from-list.interval-chart().lineWidth() governs width of interval stick #486

ds26gte committed:
interval-chart: Added .style method, takes 'bars' or 'sticks' #486.
'bars' allows setting point-size, pointer-color, lineWidth;
'sticks' only lineWidth

ds26gte committed:
interval-chart: Added 'boxes' as possible .style arg #486

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -334,7 +334,7 @@ line-width-method = method(self, lineWidth :: Number) block:
 style-method = method(self, style :: String) block:
-  when not(string-equal(style, "sticks")) and not(string-equal(style, "bars")):
+  when not(string-equal(style, "sticks")) and not(string-equal(style, "bars")) and not(string-equal(style, "boxes")):
     raise("style: must be either sticks or bars")
```
@ds26gte shouldn't this error include boxes as a recommended type?

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
+fun residual-chart-from-list(
+  xs :: P.LoN,
+  ys :: P.LoN,
+  residuals :: P.LoN
```
This shouldn't be called residuals now that we've renamed the series.

schanzer commented on src/web/arr/trove/chart.arr:
````diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
+  doc: ```
+       Consumes a list of x's, a list of y's, and a list of residuals
+       and constructs an line plot with stick intervals pointing
+       from each y to the corresponding residual
````
We should remove the "residual" text, and update accordingly

schanzer commented on src/web/arr/trove/chart.arr:
````diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
+       ```
+  xs-length = xs.length()
+  ys-length = ys.length()
+  residuals-length = residuals.length()
````
Rename

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
+  ys-length = ys.length()
+  residuals-length = residuals.length()
+  when xs-length <> ys-length:
+    raise('residual-chart: xs and ys should have the same length')
```
Rename

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
+    raise('residual-chart: xs and ys should have the same length')
+  end
+  when xs-length <> residuals-length:
+    raise('residual-chart: residuals should have the same length as xs and ys')
```
rename

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
+    raise('residual-chart: residuals should have the same length as xs and ys')
+  end
+  when xs-length == 0:
+    raise('residual-chart: need at least one datum')
```
rename

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
+  end
+  xs.each(check-num)
+  ys.each(check-num)
+  residuals.each(check-num)
```
rename

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -1873,6 +1945,38 @@ fun stacked-bar-chart-from-list(
+  xs.each(check-num)
+  ys.each(check-num)
+  residuals.each(check-num)
+  yprimes = map2(lam(y, offset): y + offset end, ys, residuals)
```
rename

schanzer commented on src/web/arr/trove/chart.arr:
```diff
@@ -2449,4 +2560,5 @@ from-list = {
   freq-bar-chart: freq-bar-chart-from-list,
   labeled-box-plot: labeled-box-plot-from-list,
   box-plot: box-plot-from-list,
+  residual-chart: residual-chart-from-list,
```
I think you've already renamed this -- now I'm wondering if Git is just showing me an old version of the file?

schanzer reviewed (changes requested):
Minor nits - I think this is good to go once the `residual` naming is gone.

ds26gte committed:
interval-chart: bad style error msg shd suggest boxes too as a good arg #486

jpolitz commented:
Want to make sure I understand the expected behavior of `interval-chart-from-list`. Here it's drawing the lines only below the dot. Also, the really small delta seems to cause the whole point to not be rendered?

![image](https://github.com/brownplt/code.pyret.org/assets/119891/57a3d956-f07d-40be-bfaf-50b247e48ab6)

jpolitz commented on src/web/js/trove/chart-lib.js:
```diff
@@ -771,6 +772,59 @@
+
+    const fillOpacity = (style == 'boxes') ? 0 : 1;
+
+    data.addRows(table.map(row => [row[0], row[1], row[1], row[2]]));
```
It's likely that the data should not be passed through uninterpreted. Probably it would be correct to call `toFixnum` on all of these if they are all numbers (and matching the columns above). e.g. https://github.com/brownplt/code.pyret.org/blob/08ad7d3e876030e0013f8b889d0d3e9d2d99d4ea/src/web/js/trove/chart-lib.js#L504

Otherwise, roughnums will cause issues.

jpolitz commented on src/web/js/trove/chart-lib.js:
```diff
@@ -771,6 +772,59 @@
+        'i0': {
+          'style': style,
+          'color': stickColor,
+          'lineWidth': stickWidth,
```
Similarly the widths and such, if they could come in as roughnums from Pyret, need to be converted somehow.

ds26gte committed:
- chart-lib.js: intervalChart, ensure all kinds of Pyret numbers accepted for
  abscissae, ordinates, point sizes, and line widths #486
- added test file test-util/pyret-programs/charts/interval-chart.arr

jpolitz commented:
Great. Thanks for the tests!
